3_optimize_ws_goroutines/kqueue_darwin.go

```diff
@@ -0,0 +1,101 @@
+// +build darwin
+
+package main
+
+import (
+	"log"
+	"reflect"
+	"sync"
+	"syscall"
+
+	"github.com/gorilla/websocket"
+)
+
+type epoll struct {
+	fd          int
+	connections map[int]*websocket.Conn
+	changes     []syscall.Kevent_t
+	lock        *sync.RWMutex
+}
+
+func MkEventsCollector() (*epoll, error) {
+	fd, err := syscall.Kqueue()
+	if err != nil {
+		return nil, err
+	}
+	_, err = syscall.Kevent(fd, []syscall.Kevent_t{{
+		Ident:  0,
+		Filter: syscall.EVFILT_USER,
+		Flags:  syscall.EV_ADD | syscall.EV_CLEAR,
+	}}, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+	return &epoll{
+		fd:          fd,
+		lock:        &sync.RWMutex{},
+		connections: make(map[int]*websocket.Conn),
+	}, nil
+}
+
+func (e *epoll) Add(conn *websocket.Conn) error {
+	fd := websocketFD(conn)
+	e.changes = append(e.changes,
+		syscall.Kevent_t{
+			Ident: uint64(fd), Flags: syscall.EV_ADD, Filter: syscall.EVFILT_READ,
+		},
+		syscall.Kevent_t{
+			Ident: uint64(fd), Flags: syscall.EV_ADD, Filter: syscall.EVFILT_WRITE,
```

Comment:
why do you need WRITE?
We only care about reading from those sockets (thats the only thing we demonstrate, reading from clients)

Reply:
I thought we need WRITE, I'll update.